sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/WatchForKafkaTopicPartitions.java

```diff
@@ -0,0 +1,184 @@
+  }
+
+  private static class ConvertToDescriptor
+      extends DoFn<KV<byte[], TopicPartition>, KafkaSourceDescriptor> {
+
+    private final @Nullable SerializableFunction<TopicPartition, Boolean> checkStopReadingFn;
+    private final @Nullable Instant startReadTime;
+    private final @Nullable Instant stopReadTime;
+
+    private ConvertToDescriptor(
+        @Nullable SerializableFunction<TopicPartition, Boolean> checkStopReadingFn,
+        @Nullable Instant startReadTime,
+        @Nullable Instant stopReadTime) {
+      this.checkStopReadingFn = checkStopReadingFn;
+      this.startReadTime = startReadTime;
+      this.stopReadTime = stopReadTime;
+    }
+
+    @ProcessElement
+    public void processElement(
+        @Element KV<byte[], TopicPartition> partition,
+        OutputReceiver<KafkaSourceDescriptor> receiver) {
+      TopicPartition topicPartition = Objects.requireNonNull(partition.getValue());
+      if (checkStopReadingFn == null || !checkStopReadingFn.apply(topicPartition)) {
+        Counter foundedTopicPartition =
+            Metrics.counter(COUNTER_NAMESPACE, topicPartition.toString());
+        foundedTopicPartition.inc();
+        receiver.output(
+            KafkaSourceDescriptor.of(
+                topicPartition, null, startReadTime, null, stopReadTime, null));
+      }
+    }
+  }
+
+  private static class WatchPartitionFn extends PollFn<byte[], TopicPartition> {
+
+    private final SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>>
+        kafkaConsumerFactoryFn;
+    private final Map<String, Object> kafkaConsumerConfig;
+    private final List<String> topics;
+
+    private WatchPartitionFn(
+        SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>> kafkaConsumerFactoryFn,
+        Map<String, Object> kafkaConsumerConfig,
+        List<String> topics) {
+      this.kafkaConsumerFactoryFn = kafkaConsumerFactoryFn;
+      this.kafkaConsumerConfig = kafkaConsumerConfig;
+      this.topics = topics;
+    }
+
+    @Override
+    public Watch.Growth.PollResult<TopicPartition> apply(byte[] element, Context c)
+        throws Exception {
+      Instant now = Instant.now();
+      return Watch.Growth.PollResult.incomplete(
+              now, getAllTopicPartitions(kafkaConsumerFactoryFn, kafkaConsumerConfig, topics))
+          .withWatermark(now);
+    }
+  }
+
+  @VisibleForTesting
+  static List<TopicPartition> getAllTopicPartitions(
+      SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>> kafkaConsumerFactoryFn,
+      Map<String, Object> kafkaConsumerConfig,
+      List<String> topics) {
+    List<TopicPartition> current = new ArrayList<>();
+    try (Consumer<byte[], byte[]> kafkaConsumer =
+        kafkaConsumerFactoryFn.apply(kafkaConsumerConfig)) {
+      if (topics != null && !topics.isEmpty()) {
+        for (String topic : topics) {
+          for (PartitionInfo partition : kafkaConsumer.partitionsFor(topic)) {
```

Comment:
What would happen if the topic does not exist? Should we handle an exception?

( I checked this : ) - https://kafka.apache.org/26/javadoc/index.html?org/apache/kafka/clients/consumer/KafkaConsumer.html)